### DISCUSSION TITLE
Adding Midi or SMPTE Sync

### DISCUSSION THREAD

**Discussion by baesek** [marcel-licence/esp32_multitrack_looper#10] [Ideas]
Hey,
i just wrote a comment under your Youtube video demonstrating the looper. I wish the looper had a sync input, midi or smpte. 
With this it could be used as a resampling device for Midi Sequencers, to record the output of a Synth and then go on creating a different part with the same Synth. I am dreaming of a pattern sampling opportunity wihout a daw, just controlled by a midi sequencer.
Do you think it is easy to implement?

regards,
paul

**Comment by marcel-licence** [comment]
Hey Paul,

MIDI is implemented to control the looper. I am not sure how to handle the sync. It could be used to get the click timing right. Playing patterns could be also possible. 
The functionality shouldn't be a big deal. The bigger problem is the limited memory (~4MB). I could imagine to crop the memory into little blocks / chunks and play them in loop or one shot samples. 

Sounds interesting to me =) 

Best regards,
Marcel

**Comment by baesek** [comment]
The mpc is sending midi timecode data every 96th note if i remember right. In a 4 bar loop it is always sending 1.1.1-to 4.4.4 and then in the 5th bar it sends 1.1.1 again. My concept so far is that the Sampler in record mode is always recording into a Buffer of 100ms and when it receives the 1.1.1 it saves a timestamp, stops buffering and records a long sample until you stop it manually by pushing a button. Then the sampler prepends the buffer to the long sample ( here some index shifting of the buffer is required so the oldest sample is at index 0). 
Now when it comes to playback there are different options:
1.  Note On playback behavior - play back the sample from note on to note off of a set note.
2. One Shot - play back the whole sample at note on
3. Midi synced playback, start playback when receicving 1.1.1 midi timecode.

Time stretch would be a future topic and is less important. I think it would be good to measure the incoming tempo while recording.
like this, on receive 1.1.1 save timestamp, then count the number of midi clocks (96th) notes until recording is stopped, check how much time has passed since start, divided by 96th notes blabla to get the tempo.
Then later on playback the incoming clocks can be analyzed every 16th note or so and compared to the original tempo while recording to find out how much faster or slower the incoming clock is.
Then perhaps there is an alogorithm which is fast enough to calculate the pseudo data to make it sound the same pitch at different speed. (I am thinking about a Akai S950 like timestretching algorithm).
If it is too complex for real time it could be also an idea to just set the Sampler into timestretch-mode and then wait for 1 bar of midi clocks, analyze it's speed, create a timestretched copy of the sample and than map this to the midi note or playback looper track instead.

In General, actually it would be amazing if there is a possibility of recording and streaming samples directly to sd-card, in the moment this is possible there are so many options with this little board. Since I desoldered the A1S with ES8388 from my audiokit board and soldered in an A1S with AC101 Codec the sound is amazing and I want to use it.
If that is too complicated, is there Any easy  way to add more RAM to the board? 
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/external-ram.html
32MB Psram? https://machdyne.com/product/qqspi-psram32/
